### DISCUSSION TITLE
Performance in cluster mode

### DISCUSSION THREAD

**Discussion by GMELUM** [uNetworking/uWebSockets.js#808] [Q&A]
Need to run a node in cluster mode to increase the vertical scaling capacity? Or will the C++ module be able to use all the cores and threads in the VPS?

**Comment by GMELUM** [comment]
> Depends on your app, if you are very cpu heavy you could run multiple uWS.js apps on different threads/processes, but there is a cost to communicate between them. Might only need a single uWS thread where adding more won't help or even hurt, limited by kernel sys calls / network / comm between threads etc

Sorry, I'm bad at English and I use a translator =)
Did I understand correctly? uWS will give me the best networking performance. And I don't need to spawn 2-3 or more uWS threads?

And to improve performance in node.js, I can use workers instead of cluster?
Thanks for your help!

**Comment by joshxyzhimself** [comment]
Worker Threads example here:

- https://github.com/uNetworking/uWebSockets.js/blob/master/examples/WorkerThreads.js

**Reply by SupertigerDev** [reply]
In socket.io, i would need to use a redis adapter to make sure all of my clients, connected to any worker instance of the website server will get the date from another instance.

How is this done? Is it done automatically?

**Comment by uNetworkingAB** [comment]
uWS is a sharp knife, but if you decide to stab yourself with it I cannot guarantee success. It's a tool not a robot.
